### PR TITLE
#11 Port from rx to coroutines

### PR DESCRIPTION
**Pull request by victorvicari**
Some dependencies were added (coroutines, test, work manager coroutines test...). Basically leaving the project structure as it was, and I moved from rx to coroutines. About the tests, I used MockK once that it's better to deal with the kotlin coroutines. Some issues came up when I updated the workmanager, I solved by putting the java `source` and `target` at the `gradle` and `kotlinOption`

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -24,17 +24,30 @@ android {
...
     // Architecture Components
     implementation("androidx.lifecycle:lifecycle-extensions:${rootProject.lifecycleVersion}")
     testImplementation("androidx.arch.core:core-testing:${rootProject.lifecycleVersion}")
     implementation("androidx.core:core-ktx:${rootProject.coreVersion}")
     implementation("androidx.fragment:fragment-ktx:${rootProject.fragmentVersion}")
     implementation("androidx.constraintlayout:constraintlayout:${rootProject.constraintLayoutVersion}")
+    // Coroutines With Arch Components
+    implementation("androidx.lifecycle:lifecycle-viewmodel-ktx:${rootProject.archCompCoroutinesVersion}")
+    implementation("androidx.lifecycle:lifecycle-runtime-ktx:${rootProject.archCompCoroutinesVersion}")
+    implementation("androidx.lifecycle:lifecycle-livedata-ktx:${rootProject.archCompCoroutinesVersion}")
```
Comment (ashdavies) [inline app/build.gradle]: Use `rootProject.lifecycleVersion` for consistency with already imported `lifecycle-extensions`, additionally using an alpha library may bring stability issues when a stable library was already in use.

File: app/build.gradle [hunk 2]
```diff
@@ -50,6 +63,8 @@ dependencies {
 
     // Mockito
     testImplementation("org.mockito:mockito-core:${rootProject.mockitoVersion}")
+    testImplementation("io.mockk:mockk:${rootProject.mockkVersion}")
```
Comment (ashdavies) [inline app/build.gradle]: If Mockito is already in use, I would really recommend using `mockito-kotlin` as it’s a bridge between the two rather than introducing a whole new mocking framework.

File: app/src/main/java/com/wtmberlin/data/ApiService.kt [hunk 3]
```diff
@@ -3,53 +3,65 @@ package com.wtmberlin.data
...
+import kotlinx.coroutines.withContext
 import org.threeten.bp.Duration
 import org.threeten.bp.Instant
 import org.threeten.bp.ZoneOffset
 import org.threeten.bp.ZonedDateTime
 
 open class ApiService(private val meetupService: MeetupService) {
-    fun events(): Single<List<WtmEvent>> {
-        return meetupService.events().map { it.map(MeetupEvent::toWtmEvent) }
+    suspend fun events() = withContext(Dispatchers.IO) {
```
Comment (ashdavies) [inline app/src/main/java/com/wtmberlin/data/ApiService.kt]: You probably shouldn’t immediately specify a dispatcher within a suspended method call, not least since it overrides the expected behaviour, but it becomes very difficult to test as the dispatcher is not injected.

File: app/src/main/java/com/wtmberlin/data/ApiService.kt [hunk 4]
```diff
@@ -3,53 +3,65 @@ package com.wtmberlin.data
...
-            append(it)
-        }
-    }.toString()
\ No newline at end of file
+            city?.let {
+                append(it)
+            }
+        }.toString()
+    }
+}
```
Comment (ashdavies) [inline app/src/main/java/com/wtmberlin/data/ApiService.kt:72]: nit: missing new line

File: app/src/main/res/layout/events_screen.xml [hunk 5]
```diff
@@ -13,10 +13,10 @@
...
 
     <androidx.swiperefreshlayout.widget.SwipeRefreshLayout
-        android:id="@+id/swipe_refresh"
-        android:layout_width="match_parent"
-        android:layout_height="match_parent"
-        app:refreshing="@{viewModel.refreshing}">
+            android:id="@+id/swipe_refresh"
+            android:layout_width="match_parent"
+            android:layout_height="match_parent"
+            app:refreshing="@{safeUnbox(viewModel.refreshing)}">
```
Comment (ashdavies) [inline app/src/main/res/layout/events_screen.xml]: Does this project use `safeUnbox`? This seems unrelated to this PR.

File: app/src/test/java/com/wtmberlin/ui/EventsDetailsViewModelTest.kt [hunk 6]
```diff
@@ -1,107 +1,162 @@
...
 
 import androidx.arch.core.executor.testing.InstantTaskExecutorRule
-import com.google.common.truth.Truth.assertThat
-import com.wtmberlin.*
-import com.wtmberlin.data.Coordinates
-import com.wtmberlin.data.Repository
-import com.wtmberlin.data.Result
-import com.wtmberlin.data.WtmEvent
-import io.reactivex.processors.PublishProcessor
+import com.wtmberlin.data.*
```
Comment (ashdavies) [inline app/src/test/java/com/wtmberlin/ui/EventsDetailsViewModelTest.kt]: nit: avoid wildcard imports

File: app/src/test/java/com/wtmberlin/ui/EventsDetailsViewModelTest.kt [hunk 7]
```diff
@@ -1,107 +1,162 @@
...
     @get:Rule
     val instantTaskExecutorRule = InstantTaskExecutorRule()
 
-    val mockEvent = PublishProcessor.create<Result<WtmEvent>>()
-    val mockRepository = mock<Repository>().also {
-        `when`(it.event("1")).thenReturn(mockEvent)
-    }
+    @MockK
+    internal lateinit var repository: Repository
+    private val viewModel by lazy { EventDetailsViewModel("1", repository) }
```
Comment (ashdavies) [inline app/src/test/java/com/wtmberlin/ui/EventsDetailsViewModelTest.kt]: Use `LazyThreadSafetyMode.NONE` with the `lazy` delegate to avoid having the performance cost of a synchronised double check when accessing in tests or in single threaded environments.

File: app/src/test/java/com/wtmberlin/works/RefreshEventsJobTest.kt [hunk 8]
```diff
@@ -1,67 +1,80 @@
...
+    @MockK
+    private lateinit var notifications: Notifications
+
+    private val work by lazy { RefreshEventsWorker(ctx, params, apiService, eventDao, notifications) }
+    private val testDispatcher = TestCoroutineDispatcher()
+
+    @Before
+    fun setUp() {
+        MockKAnnotations.init(this)
+        Dispatchers.setMain(testDispatcher)
```
Comment (ashdavies) [inline app/src/test/java/com/wtmberlin/works/RefreshEventsJobTest.kt]: Consider creating a `Rule` using a `TestDispatcher` where you don't need to manipulate it under test

File: app/src/test/java/com/wtmberlin/works/RefreshEventsJobTest.kt [hunk 9]
```diff
@@ -1,67 +1,80 @@
...
+        val eventApi1 = eventDb.copy(id = "15")
+        val eventApi2 = eventDb.copy(id = "14")
+        val listApi = listOf(eventApi1, eventApi2)
+        coEvery { eventDao.getAll() } returns listOf(eventDb)
+        coEvery { apiService.events() } returns listApi
 
-        worker.doWork()
+        work.doWork()
 
-        verify(eventDao).replaceAll(listOf(eventFromNetwork))
```
Comment (ashdavies) [inline app/src/test/java/com/wtmberlin/works/RefreshEventsJobTest.kt:44]: It really doesn’t seem that necessary to migrate the mocking framework to something else when `Mockito` is already in use.

**Review by ashdavies — CHANGES_REQUESTED** [review]
Thanks for the contribution @victorvicari , I have a few comments regarding your pull request.

- Verify that chaning the JVM target from 7 to 8 doesn’t introducing any regression issues
- Inject `Dispatcher`’s instead of hard coding, and only apply then inside a coroutine context
- I don’t think it’s appropriate to introduce a new mock framework when one is already in use
- Use lifecycle `2.1.0-rc01` instead of alpha to make use of `viewModelScope` instead of alpha

You can find a good example of Retrofit, Coroutine and Dispatcher use with the Plaid project.
https://github.com/android/plaid/blob/885a002a4f7f56ed7ff8c69c7866957591727a4c/core/src/main/java/io/plaidapp/core/data/CoroutinesDispatcherProvider.kt

File: app/src/test/java/com/wtmberlin/ui/EventsViewModelTest.kt [hunk 10]
```diff
@@ -1,92 +1,126 @@
...
-    private val simpleUpcomingEvent = mockk<WtmEvent>().also {
-        val dateStart = LocalDateTime.now().plusDays(1)
-        every { it.id } returns "1"
-        every { it.dateTimeStart } returns ZonedDateTime.of(dateStart, ZoneId.systemDefault())
-        every { it.name } returns "nameEvent"
-        every { it.venue } returns null
-        every { it.meetupUrl } returns "www.google.com"
-    }
+    private fun startEvents(event: WtmEvent = mock()): WtmEvent {
+        Mockito.`when`(event.id).thenReturn("1")
```
Comment (ashdavies) [inline app/src/test/java/com/wtmberlin/ui/EventsViewModelTest.kt]: You can also use the `given` … `when` … `then` syntax from `BDDMockito` to avoid protected keywords in Kotlin

**Review by ashdavies — APPROVED** [review]
Great improvement! Thanks for putting the effort in  👍 💯

File: app/build.gradle (hunk 1 above)
Comment (victorvicari) [reply app/build.gradle]: Ok.

File: app/build.gradle (hunk 2 above)
Comment (victorvicari) [reply app/build.gradle]: I understand

**Comment by victorvicari** [comment]
Still missing some things. I'll try to finish today.

File: app/src/test/java/com/wtmberlin/ui/EventsViewModelTest.kt (hunk 10 above)
Comment (victorvicari) [reply app/src/test/java/com/wtmberlin/ui/EventsViewModelTest.kt]: Done. Thanks for the tip

File: app/src/test/java/com/wtmberlin/works/RefreshEventsJobTest.kt (hunk 8 above)
Comment (victorvicari) [reply app/src/test/java/com/wtmberlin/works/RefreshEventsJobTest.kt]: Done

File: app/src/test/java/com/wtmberlin/ui/EventsDetailsViewModelTest.kt (hunk 7 above)
Comment (victorvicari) [reply app/src/test/java/com/wtmberlin/ui/EventsDetailsViewModelTest.kt]: Ok

File: app/src/test/java/com/wtmberlin/SetMainDispatcherRule.kt [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+import kotlinx.coroutines.test.resetMain
+import kotlinx.coroutines.test.setMain
+import org.junit.rules.TestWatcher
+import org.junit.runner.Description
+
+@ExperimentalCoroutinesApi
+class SetMainDispatcherRule : TestWatcher() {
+
+    override fun starting(description: Description?) {
+        Dispatchers.setMain(newSingleThreadContext("UI thread"))
```
Comment (ashdavies) [inline app/src/test/java/com/wtmberlin/SetMainDispatcherRule.kt:15]: Don’t forget you’ll need to release this thread on test finish

Comment (faogustavo) [reply app/src/test/java/com/wtmberlin/SetMainDispatcherRule.kt:15]: But he is already doing this

File: app/build.gradle [hunk 12]
```diff
@@ -43,13 +55,20 @@ dependencies {
...
     implementation("org.koin:koin-android-viewmodel:${rootProject.koinVersion}")
 
     // Mockito
     testImplementation("org.mockito:mockito-core:${rootProject.mockitoVersion}")
+    implementation("org.jetbrains.kotlinx:kotlinx-coroutines-core:${rootProject.coroutinesTestCoreVersion}")
+    testImplementation("org.jetbrains.kotlinx:kotlinx-coroutines-core:${rootProject.coroutinesTestCoreVersion}")
+    implementation("org.jetbrains.kotlinx:kotlinx-coroutines-android:${rootProject.coroutinesTestCoreVersion}")
+    testImplementation("org.jetbrains.kotlinx:kotlinx-coroutines-android:${rootProject.coroutinesTestCoreVersion}")
+    testImplementation("org.jetbrains.kotlinx:kotlinx-coroutines-test:${rootProject.coroutinesTestVersion}")
+    testImplementation("com.nhaarman.mockitokotlin2:mockito-kotlin:${rootProject.mockitoKotlinVersion}")
```
Comment (faogustavo) [inline app/build.gradle]: Can you split the `implementation` from `testImplementation` to make it a little bit more organized

File: app/src/main/java/com/wtmberlin/data/Repository.kt [hunk 13]
```diff
@@ -1,67 +1,56 @@
...
-                }
-            }
-            .doOnSubscribe { eventsResource.refresh() }
+    suspend fun events() = withContext(dispatchers.io) {
+        fetchEvents()
     }
 
-    fun refreshEvents() {
-        eventsResource.refresh()
+    suspend fun refreshEvents() = withContext(dispatchers.io) {
```
Comment (faogustavo) [inline app/src/main/java/com/wtmberlin/data/Repository.kt:26]: You can simplify it by doing this
```suggestion
    suspend fun refreshEvents() = withContext(dispatchers.io, ::fetchEvents)
```

File: app/src/main/java/com/wtmberlin/data/WtmEventDao.kt [hunk 14]
```diff
@@ -4,15 +4,14 @@ import androidx.room.Dao
...
 
 @Dao
 interface WtmEventDao {
     @Query("SELECT * FROM WtmEvent")
-    fun getAll(): Flowable<List<WtmEvent>>
+    suspend fun getAll(): List<WtmEvent>
 
     @Query("SELECT * FROM WtmEvent WHERE id = :eventId")
-    fun getById(eventId: String): Flowable<WtmEvent>
+    suspend fun getById(eventId: String): WtmEvent
```
Comment (faogustavo) [inline app/src/main/java/com/wtmberlin/data/WtmEventDao.kt]: You should use an optional here, to avoid crashing the app if the item doesn't exist.
```suggestion
    suspend fun getById(eventId: String): WtmEvent?
```

File: app/src/main/java/com/wtmberlin/util/ErrorLogger.kt [hunk 15]
```diff
@@ -0,0 +1,10 @@
+package com.wtmberlin.util
+
+class ErrorLogger {
+
+    fun getException(e: Throwable) {
```
Comment (faogustavo) [inline app/src/main/java/com/wtmberlin/util/ErrorLogger.kt]: Maybe you can choose a meaningful name, like `logException`

File: app/src/main/java/com/wtmberlin/util/ErrorLogger.kt [hunk 16]
```diff
@@ -0,0 +1,10 @@
...
+        //Todo: add a crash lib
+        val a = e
```
Comment (faogustavo) [inline app/src/main/java/com/wtmberlin/util/ErrorLogger.kt]: I think is missing the Timber rule here. It will be nice to also check if its a prod build and avoid logging with timber and then, maybe, add a crash lib like `crashlytics` or `sentry`.

File: app/src/test/java/com/wtmberlin/ui/EventsDetailsViewModelTest.kt [hunk 17]
```diff
@@ -1,107 +1,166 @@
...
-        )
+        verify(mockErrorLogger, times(1)).getException(any())
+    }
+
+    @ExperimentalCoroutinesApi
+    private fun unconfinifyTestScope() {
+        ui = Dispatchers.Unconfined
+        io = Dispatchers.Unconfined
+        background = Dispatchers.Unconfined
     }
```
Comment (faogustavo) [inline app/src/test/java/com/wtmberlin/ui/EventsDetailsViewModelTest.kt]: Can you create a rule with this? Like the `SetMainDispatcherRule`?

File: build.gradle [hunk 18]
```diff
@@ -12,24 +12,28 @@ buildscript {
     ext.gradleVersion = '3.3.1'
     ext.junitVersion = "4.12"
     ext.koinVersion = "1.0.2"
-    ext.kotlinVersion = "1.3.21"
-    ext.lifecycleVersion = "2.0.0"
+    ext.kotlinVersion = "1.3.41"
+    ext.lifecycleVersion = "2.1.0-rc01"
```
Comment (faogustavo) [inline build.gradle:16]: Do you really need the `-rc` version?

**Review by faogustavo — APPROVED** [review]
I will be nice to create a new issue to migrate all dependencies to a `buildSrc` "module". https://handstandsam.com/2018/02/11/kotlin-buildsrc-for-better-gradle-dependency-management/

File: app/src/main/java/com/wtmberlin/data/WtmEventDao.kt (hunk 14 above)
Comment (victorvicari) [reply app/src/main/java/com/wtmberlin/data/WtmEventDao.kt]: Ok. I'll add

File: app/src/test/java/com/wtmberlin/ui/EventsDetailsViewModelTest.kt (hunk 17 above)
Comment (victorvicari) [reply app/src/test/java/com/wtmberlin/ui/EventsDetailsViewModelTest.kt]: Done

File: build.gradle (hunk 18 above)
Comment (victorvicari) [reply build.gradle:16]: It was one of the @ashdavies suggestions

File: app/build.gradle (hunk 12 above)
Comment (victorvicari) [reply app/build.gradle]: Done

File: app/src/main/java/com/wtmberlin/util/ErrorLogger.kt (hunk 15 above)
Comment (victorvicari) [reply app/src/main/java/com/wtmberlin/util/ErrorLogger.kt]: Done

File: app/src/main/java/com/wtmberlin/util/ErrorLogger.kt (hunk 16 above)
Comment (victorvicari) [reply app/src/main/java/com/wtmberlin/util/ErrorLogger.kt]: What do you think @ashdavies

Comment (ashdavies) [reply app/src/main/java/com/wtmberlin/util/ErrorLogger.kt]: Conditional reporting can be achieved with a debug `Timber.Tree` which can always be beneficial, but crash reporting should be configured by the project owner since it requires setup of an additional service and an API key which can sometimes incur billing.

**Comment by annamorgiel** [comment]
Hey @victorvicari @faogustavo thanks so much for the input 🌷 How should we process with this one? We could rebase the branch to master and resolve the issues, a fresh PR might be a good idea.

File: app/src/test/java/com/wtmberlin/SetMainDispatcherRule.kt (hunk 11 above)
Comment (ashdavies) [reply app/src/test/java/com/wtmberlin/SetMainDispatcherRule.kt:15]: https://github.com/Kotlin/kotlinx.coroutines/tree/master/kotlinx-coroutines-test

Check the example where the surrogate thread is being closed.

File: app/src/main/java/com/wtmberlin/util/CoroutineViewModel.kt [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlin.coroutines.CoroutineContext
+
+var ui: CoroutineDispatcher = Dispatchers.Main
+var io: CoroutineDispatcher = Dispatchers.IO
+var background: CoroutineDispatcher = Dispatchers.Default
+
+abstract class CoroutineViewModel : ViewModel(), CoroutineScope {
```
Comment (ashdavies) [inline app/src/main/java/com/wtmberlin/util/CoroutineViewModel.kt:14]: This implementation is already available with `viewModelScope` in `lifecycle:2.1.0-RC`

**Comment by annamorgiel** [comment]
@victorvicari if you plan to merge the other PR, can this one be closed?

**Comment by victor-vicari-codeminer42** [comment]
Sure thing.

**Comment by annamorgiel** [comment]
Closing because update version was merged in #56. Big thanks to everyone!
